### ISSUE TITLE
Test the `panacead migrate` command

### ISSUE THREAD

**Issue by youngjoon-lee** [medibloc/panacea-core#154]
As described in the [cosmoshub upgrade guide](https://github.com/cosmos/gaia/blob/main/docs/migration/cosmoshub-3.md#upgrade-procedure), we probably need to run the `panacead migrate` command to migrate an exported genesis file from `panacea-2` to the genesis file for `panacea-3`.
The `panacead migrate` command may already have some migration logic implemented by cosmos-sdk because the genesis file format might have changed in the Stargate.

Thus, let's check if the current `panacead migrate` command in the `stargate` branch is enough for our upgrade scenario. If not, add more migration logic.

**Comment by youngjoon-lee** [comment]
When I tried to run the `panacead migrate ...`, I got this error:

> Error: error reading GenesisDoc at /home/oudwud/panacea-2-export.json: evidence.MaxAgeNumBlocks must be greater than 0. Got 0. Make sure that you have correctly migrated all Tendermint consensus params, please see the chain migration guide at https://docs.cosmos.network/master/migrations/chain-upgrade-guide-040.html for more info

It seems that we have to migrate some Tendermint params manually. Let me check.

**Comment by youngjoon-lee** [comment]
> It seems that we have to migrate some Tendermint params manually.

We can migrate Tendermint consensus params manually as below before running `panacea migrate ...`.
```bash
cat ~/panacea-2-export.json | \
jq -c '.consensus_params.evidence.max_age_duration = "1814400000000000"' | \
jq -c '.consensus_params.evidence.max_age_num_blocks = "1814400"' | \
jq -c '.consensus_params.evidence.max_bytes = "50000"' | \
jq -c 'del(.consensus_params.evidence.max_age)' | \
jq > ~/panacea-2-export-manually-migrated.json
```
We've already noticed about this: https://github.com/medibloc/panacea-core/pull/126#discussion_r642280864.

Cosmos Hub uses the following params which are slightly different from ours, but I think we can use our settings above (To be honest, I don't exactly know what the evidence is. I'll investigate it soon).
```json
    "evidence": {
      "max_age_duration": "172800000000000",
      "max_age_num_blocks": "1000000",
      "max_bytes": "50000"
    },
```

**Comment by youngjoon-lee** [comment]
After migrating the genesis file successfully, I tried to run the `panacead start`. But, it crashed with the following panic:
```
panic: invariant broken: bank: total supply invariant
        sum of accounts coins: 1000000000001umed
        supply.Total:          100000128337069umed
        CRITICAL please submit the following transaction:
                 tx crisis invariant-broken bank total-supply       
```
When panacead is instantiated, it checks whether all values in the genesis are 'reasonable'. This procedure called as 'invariants' checks.

It's weird that it returned the `total supply invariant`. I already saw that sum of accounts coins in the genesis is 100000128337069umed, not 1000000000001umed.
I will check if this is a bug from cosmos-sdk.

**Comment by gyuguen** [comment]
> > It seems that we have to migrate some Tendermint params manually.
> 
> We can migrate Tendermint consensus params manually as below before running `panacea migrate ...`.
> 
> ```shell
> cat ~/panacea-2-export.json | \
> jq -c '.consensus_params.evidence.max_age_duration = "1814400000000000"' | \
> jq -c '.consensus_params.evidence.max_age_num_blocks = "1814400"' | \
> jq -c '.consensus_params.evidence.max_bytes = "50000"' | \
> jq -c 'del(.consensus_params.evidence.max_age)' | \
> jq > ~/panacea-2-export-manually-migrated.json
> ```
> 
> I'll update our upgrade instruction in the `panacea-launch` and open a PR.
> 
> We've already noticed about this: [#126 (comment)](https://github.com/medibloc/panacea-core/pull/126#discussion_r642280864).
> 
> Cosmos Hub uses the following params which are slightly different from ours, but I think we can use our settings above (To be honest, I don't exactly know what the evidence is. I'll investigate it soon).
> 
> ```json
>     "evidence": {
>       "max_age_duration": "172800000000000",
>       "max_age_num_blocks": "1000000",
>       "max_bytes": "50000"
>     },
> ```

Umm.. It's not easy to find evidence.
It is maybe associated of the consensus.
I think it need to study more.T_T

**Comment by youngjoon-lee** [comment]
> After migrating the genesis file successfully, I tried to run the panacead start. But, it crashed with the following panic:

I resolved that.
The all account balances have been moved from `.app_state.accounts` to `.bank.balances` gradually through the history of cosmos-sdk v0.37 ~ v0.40.

In order to apply all migrations from v0.37 to v0.40, we must run all migration commands as below:
```bash
panacead migrate v0.38 ~/panacea-2-export-jq-cs.json --chain-id panacea-3 | jq > 38.json
panacead migrate v0.39 ./38.json --chain-id panacea-3 | jq > 39.json
panacead migrate v0.40 ./39.json --chain-id panacea-3 | jq > 40.json
```
Then, the `bank` invariant check is passed.

But, I got another invariant check failure:
```
panic: unknown field "total_topics" in types.Owner
```
Let me investigate it.

**Comment by youngjoon-lee** [comment]
> But, I got another panic (which isn't from invariant checks):

I found the reason.
So far, the genesis is exported with snake-case:
```json
{
  "app_hash": "", 
  "app_state": {
    "aol": {
      "owners": {
        "panacea1cd4xv9am7xqvwxkcry9z4zdnj9vxkfwk6d5rdz": {
          "total_topics": "1" 
        }
      }
```
But, proto json marshaller expects camel-case: https://github.com/medibloc/panacea-core/blob/8f8223bd73d6d1aabb9eb0a56f1f402ff7c49f47/x/aol/types/owner.pb.go#L27

Well. Let's make proto accepts snake-case instead of camel-case, because all other json fields are written in snake-case:
https://github.com/cosmos/cosmos-sdk/blob/a7e19ad4588bcceeeeab522f5807b9ff22e91a8c/x/bank/types/bank.pb.go#L31

However, the only one exceptions is the `x/did`. The `x/did` genesis is intentionally exported with camel-case in order to obey the W3C spec. So, let's leave it as it is.
```json
    "did": {
      "documents": {
        "did:panacea:2SXdL1tLWfjV4SoEbBHXXEQK3vrfEgsoajgS78GgBst8": {
          "document": {
            "@context": "https://www.w3.org/ns/did/v1",
            "authentication": [
              "did:panacea:2SXdL1tLWfjV4SoEbBHXXEQK3vrfEgsoajgS78GgBst8#key1"
            ],
            "id": "did:panacea:2SXdL1tLWfjV4SoEbBHXXEQK3vrfEgsoajgS78GgBst8",
            "verificationMethod": [
              {
                "controller": "did:panacea:2SXdL1tLWfjV4SoEbBHXXEQK3vrfEgsoajgS78GgBst8",
                "id": "did:panacea:2SXdL1tLWfjV4SoEbBHXXEQK3vrfEgsoajgS78GgBst8#key1",
                "publicKeyBase58": "jkfr6i8vtpDNhBWxDnLnJtwjNi7foGTXQKFiyoQSt5Ka",
                "type": "Secp256k1VerificationKey2018"
              }
            ]
          },
          "sequence": "0"
        },
```

**Comment by gyuguen** [comment]
The following error occurs when registering the Token module.

```
panic: invariant broken: bank: total supply invariant
	sum of accounts coins: 10000000000000ugyu993,1000000456309256umed
	supply.Total:          1000000456309256umed


	CRITICAL please submit the following transaction:
		 tx crisis invariant-broken bank total-supply
```

I'm looking for the cause.

**Comment by youngjoon-lee** [comment]
> The following error occurs when registering the Token module.

Ah. It seems that the `ugyu993` token wasn't added to the total supply. It might be a bug of the `x/token`, or `panacead migrate ...`.

**Comment by gyuguen** [comment]
> > The following error occurs when registering the Token module.
> 
> Ah. It seems that the `ugyu993` token wasn't added to the total supply. It might be a bug of the `x/token`, or `panacead migrate ...`.

Yes, the cause is `genesis.json` bank.supply of 'x/token' was not added to supply.
Looking at the export file in v0.36, it was not added to the supply.

I think we'd better handle this manually.

**Comment by gyuguen** [comment]
Or maybe it's better to delete the token since it's not properly opened yet.

**Comment by youngjoon-lee** [comment]
> Yes, the cause is genesis.json bank.supply of x/token was not added to supply.
Looking at the export file in v0.36, it was not added to the supply.

You're right. I remember that Panacea v1.3.3 has that bug. Let me check if any token has been created in the mainnet. If so, we can release `v1.3.3-hotfix.1` and upgrade the mainnet with two steps (v1.3.3 -> v1.3.3-hotfix.1 -> export -> v2.0.0).

**Comment by youngjoon-lee** [comment]
> > Yes, the cause is genesis.json bank.supply of x/token was not added to supply.
> 
> Looking at the export file in v0.36, it was not added to the supply.
> 
> 
> 
> You're right. I remember that Panacea v1.3.3 has that bug. Let me check if any token has been created in the mainnet. If so, we can release `v1.3.3-hotfix.1` and upgrade the mainnet with two steps (v1.3.3 -> v1.3.3-hotfix.1 -> export -> v2.0.0).

Ah, no. We can't. 
We need to modify the genesis file manually as you said. 
Let me check if the mainnet has any tokens.

**Comment by gyuguen** [comment]
OK. I understand.
I will continue the test.

**Comment by youngjoon-lee** [comment]
This issue was closed because the PR #166 was merged. But, please reopen this issue whenever you want.
